Add test for distroless docker image #192

JonoYang opened the pull request:
This PR adds a Docker pipeline test for a test Docker image based on docker://gcr.io/distroless/base

JonoYang commented:
@tdruez The new test I added is failing due to a difference in how the file results in the JSON output are ordered. I modified the queryset used to get the `CodebaseResources` for the JSON output such that the `CodebaseResources` are ordered by path. However, that did not seem to help with the problem. Have you encountered this before?

tdruez commented:
>  I modified the queryset used to get the CodebaseResources for the JSON output such that the CodebaseResources are ordered by path

The CodebaseResources should already be ordered by path based on the model `ordering = ("project", "path")`.

> Have you encountered this before?

I don't recall dealing with results ordering recently.

JonoYang commented:
I just find it very strange that I am getting different result ordering when running the pipeline tests on my Ubuntu 18.04 laptop, Ubuntu 20.04 WSL2 instance, and a Ubuntu 20.04 LXC container.

tdruez commented:
> I just find it very strange that I am getting different result ordering when running the pipeline tests on my Ubuntu 18.04 laptop, Ubuntu 20.04 WSL2 instance, and a Ubuntu 20.04 LXC container.

@JonoYang let's try to locate the initial place where the order differs:
- During the initial listing of the files on disk by container_inspector?
or 
- During the serialization of the data from the DB to a results file?

This will help to fix the issue.

JonoYang commented:
@tdruez 

The ordering differs when we are iterating through the layer resources (https://github.com/nexB/scancode.io/blob/6c4ba7bdadf2c9f64b012c17f7b49e4c98ac7d17/scanpipe/pipes/docker.py#L133) and when we collect the CodebaseResources for the JSON output (https://github.com/nexB/scancode.io/blob/192-add-distroless-test/scanpipe/pipes/output.py#L189) I think the issue is how postgres or the django orm is ordering the Resources for the JSON output.

I captured the paths of the Resource collected from my laptop running Ubuntu 18.04 and my Ubuntu 20.04 instance running in a Linux LXC container.

[jsonoutput-paths-ubuntu-18.04.txt](https://github.com/nexB/scancode.io/files/8944695/jsonoutput-paths-ubuntu-18.04.txt)
[jsonoutput-paths-ubuntu-20.04.txt](https://github.com/nexB/scancode.io/files/8944694/jsonoutput-paths-ubuntu-20.04.txt)

The difference in ordering is immediately shown at the top of the two files. 

These are the lines 4, 5, and 6 of `jsonoutput-paths-ubuntu-18.04.txt`:
```
gcr_io_distroless_base.tar.gz-extract/cb3279093e638ddfd56bff4d3d89c5a3ed6dd59dbcfbc2f3107045635996b822/lib/x86_64-linux-gnu/libanl-2.31.so
gcr_io_distroless_base.tar.gz-extract/cb3279093e638ddfd56bff4d3d89c5a3ed6dd59dbcfbc2f3107045635996b822/lib/x86_64-linux-gnu/libBrokenLocale-2.31.so
gcr_io_distroless_base.tar.gz-extract/cb3279093e638ddfd56bff4d3d89c5a3ed6dd59dbcfbc2f3107045635996b822/lib/x86_64-linux-gnu/libc-2.31.so
```

 However, lines 4, 5, and 6 of `jsonoutput-paths-ubuntu-20.04.txt` are:
 ```
 gcr_io_distroless_base.tar.gz-extract/cb3279093e638ddfd56bff4d3d89c5a3ed6dd59dbcfbc2f3107045635996b822/lib/x86_64-linux-gnu/libBrokenLocale-2.31.so
gcr_io_distroless_base.tar.gz-extract/cb3279093e638ddfd56bff4d3d89c5a3ed6dd59dbcfbc2f3107045635996b822/lib/x86_64-linux-gnu/libSegFault.so
gcr_io_distroless_base.tar.gz-extract/cb3279093e638ddfd56bff4d3d89c5a3ed6dd59dbcfbc2f3107045635996b822/lib/x86_64-linux-gnu/libanl-2.31.so
 ```
 
 My initial guess is that the paths from my Ubuntu 20.04 instance are being ordered by ASCII value, rather than lexicographical order.

tdruez commented:
Thanks for the investigation, I've seen similar issue in the past.

> I think the issue is how postgres or the django orm is ordering the Resources for the JSON output.

It could also be at the OS level, PostgreSQL using some sorting logic from the OS setup.

See https://stackoverflow.com/questions/34537982/incorrect-sort-collation-order-with-spaces-in-postgresql-9-4 and https://unix.stackexchange.com/questions/252419/unexpected-sort-order-in-en-us-utf-8-locale/252426#252426

Could you check the database "Encoding", "Collate", and "Ctype" on each of your instances?

```
$ ./manage.py dbshell
scancodeio=> \l
```

```
                                   List of databases
    Name    |   Owner    | Encoding |   Collate   |    Ctype    |   Access privileges
------------+------------+----------+-------------+-------------+-----------------------
 scancodeio | scancodeio | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
```

Let me know if they all share the same setup.

JonoYang commented:
@tdruez 

Thanks for the resources on the DB encoding! They're very useful. I checked the DB and collation settings on my postgres instances and on my Ubuntu 18.04 laptop, and it is `en_US.UTF-8`:
```
                                   List of databases
    Name    |   Owner    | Encoding |   Collate   |    Ctype    |   Access privileges   
------------+------------+----------+-------------+-------------+-----------------------
 postgres   | postgres   | UTF8     | en_US.UTF-8 | en_US.UTF-8 | 
 scancodeio | scancodeio | UTF8     | en_US.UTF-8 | en_US.UTF-8 | 
 template0  | postgres   | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
            |            |          |             |             | postgres=CTc/postgres
 template1  | postgres   | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
            |            |          |             |             | postgres=CTc/postgres
```

On my Ubuntu 20.04 instance, the encoding is different:
```
                               List of databases
    Name    |   Owner    | Encoding | Collate |  Ctype  |   Access privileges   
------------+------------+----------+---------+---------+-----------------------
 postgres   | postgres   | UTF8     | C.UTF-8 | C.UTF-8 | 
 scancodeio | scancodeio | UTF8     | C.UTF-8 | C.UTF-8 | 
 template0  | postgres   | UTF8     | C.UTF-8 | C.UTF-8 | =c/postgres          +
            |            |          |         |         | postgres=CTc/postgres
 template1  | postgres   | UTF8     | C.UTF-8 | C.UTF-8 | =c/postgres          +
            |            |          |         |         | postgres=CTc/postgres
```

In `ci.yml`, I've added `POSTGRES_INITDB_ARGS` under `env` on the postgres settings, where I set `POSTGRES_INITDB_ARGS` to `--encoding=UTF-8 --lc-collate=en_US.UTF-8 --lc-ctype=en_US.UTF-8` to ensure that the database encoding is `en_US.UTF-8`

I've modified `ci.yml` with an extra step to show the DB encoding:
```
Run source bin/activate; DJANGO_SETTINGS_MODULE=scancodeio.settings django-admin dbshell -- -c "\l"
  
                                     List of databases
    Name    |   Owner    | Encoding |   Collate   |    Ctype    |     Access privileges     
------------+------------+----------+-------------+-------------+---------------------------
 postgres   | scancodeio | UTF8     | en_US.UTF-8 | en_US.UTF-8 | 
 scancodeio | scancodeio | UTF8     | en_US.UTF-8 | en_US.UTF-8 | 
 template0  | scancodeio | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/scancodeio            +
            |            |          |             |             | scancodeio=CTc/scancodeio
 template1  | scancodeio | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/scancodeio            +
            |            |          |             |             | scancodeio=CTc/scancodeio
```

The encoding on the postgres instance set up by github actions is the same as on my laptop, but the resource order is still different.

JonoYang commented:
From https://stackoverflow.com/questions/13370051/ordering-differences-between-postgres-instances-on-different-machines-same-loca , the difference in Resource ordering may be an OS issue.

I've run the test provided by the answer:
```
$ LC_CTYPE=en_US.UTF-8 sort << EOF
> i
> N
> EOF
```

On my Ubuntu 18.04 laptop, the results are:
```
i
N
```

On my Ubuntu 20.04 instance:
```
N
i
```

On the GitHub actions CI (https://github.com/nexB/scancode.io/runs/6992076076?check_suite_focus=true#step:9:1):
```
N
i
```

tdruez commented:
macOS 12.4 returns the same ordering as Ubuntu 20.04 and the CI.

> This test still fails on Ubuntu 18.04.

Since this is OS related and we cannot do much in the ScanCode.io to solve this, I'm fine with the issue on Ubuntu 18 since this is not the env we officially support for ScanCode.io.

Did you find a way to force Ubuntu 18 to behave like Ubuntu 20 regarding the sorting?

Also, on Ubuntu 18.04, are you running the code directly from the source or do run through Docker? I'm wondering if the host OS (Ubuntu 18) would have an impact on the sorting that takes place in the Docker container OS (Ubuntu 20).

tdruez commented:
@JonoYang I've added some changes to your branch to ensure we have the same PostgreSQL setup everywhere: Docker, CI, and Makefile.
See https://github.com/nexB/scancode.io/commit/af93c1c7f2f968a004521508ac2c1141b678a284

JonoYang commented:
@tdruez 

I haven't found a way to get the sorting on Ubuntu 18.04 to work like it does on 20.04. I reinitialized the database using `en_US.UTF-8` for collation and ctype, but I'm still failing the distroless docker image test. I'll deal with that for now, but I'll update my laptop to Ubuntu 22.04 or Fedora 36 when I have some free time.

tdruez commented:
@JonoYang the output for the distroless test is kind of huge for a test. Is there any way to make the input content much smaller? I'm concerned that this will be a pain to maintain  in the future (the diffs are too big).
